### PR TITLE
ECC correction of VPD

### PR DESCRIPTION
**Pull request by GiridhariKrishnan**
ECC corrects VPD data in 32:1 bit ratio in the
event of corruption.

### CODE REVIEW

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
This PR has merge conflicts, please fix that.

**Comment by GiridhariKrishnan** [comment]
> This PR has merge conflicts, please fix that.

Yes, that is due to updation of base 1050 branch, fixing that.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
The commit header is misleading. It appears as if it corrects ECC of VPD.

File: vpd-manager/editor_impl.cpp [hunk 1]
```diff
@@ -214,6 +214,18 @@ auto EditorImpl::getValue(offsets::Offsets offset)
     return lowByte;
 }
 
+void EditorImpl::checkRecordECC()
+{
+    auto itrToRecordData = vpdFile.cbegin();
+    std::advance(itrToRecordData, thisRecord.recOffset);
```
Comment (jinuthomas) [inline vpd-manager/editor_impl.cpp:220]: check offset if need to be one less or not

File: vpd-manager/editor_impl.cpp [hunk 2]
```diff
@@ -214,6 +214,18 @@ auto EditorImpl::getValue(offsets::Offsets offset)
...
+
+    auto itrToRecordECC = vpdFile.cbegin();
+    std::advance(itrToRecordECC, thisRecord.recECCoffset);
```
Comment (jinuthomas) [inline vpd-manager/editor_impl.cpp:223]: same as above

File: vpd-manager/editor_impl.cpp [hunk 3]
```diff
@@ -624,6 +653,9 @@ void EditorImpl::updateKeyword(const Binary& kwdData, uint32_t offset,
 
                 // update the data to the file
                 updateData(kwdData);
+            
+               // Check Data before updating
+               checkRecordECC();
```
Comment (jinuthomas) [inline vpd-manager/editor_impl.cpp:658]: lets check this in demo

File: vpd-manager/editor_impl.hpp [hunk 4]
```diff
@@ -171,6 +171,9 @@ class EditorImpl
     void makeDbusCall(const std::string& object, const std::string& interface,
                       const std::string& property, const std::variant<T>& data);
 
+    /** @brief Method to check the record's Data before updating */
```
Comment (jinuthomas) [inline vpd-manager/editor_impl.hpp:174]: better brief if possible

File: vpd-manager/manager.cpp [hunk 5]
```diff
@@ -168,7 +168,8 @@ void Manager::restoreSystemVpd()
             jsonFile["frus"][systemVpdFilePath][0]["inventoryPath"]
                 .get_ref<const nlohmann::json::string_t&>();
 
-        parser = ParserFactory::getParser(vpdVector, (pimPath + inventoryPath));
+        parser = ParserFactory::getParser(vpdVector, (pimPath + inventoryPath),
+                                          systemVpdFilePath, 0);
```
Comment (jinuthomas) [inline vpd-manager/manager.cpp:172]: No magic numbers

File: vpd-parser/ipz_parser.hpp [hunk 6]
```diff
@@ -63,6 +64,13 @@ class IpzVpdParser : public ParserInterface
 
     /*Inventory path of the FRU */
     const std::string inventoryPath;
+
+    /* VPD Path */
+    const std::string vpdPath;
```
Comment (jinuthomas) [inline vpd-parser/ipz_parser.hpp:69]: vpdFilePath

File: vpd-parser/ipz_parser.hpp [hunk 7]
```diff
@@ -63,6 +64,13 @@ class IpzVpdParser : public ParserInterface
...
+
+    /* Offset */
+    uint32_t offset;
```
Comment (jinuthomas) [inline vpd-parser/ipz_parser.hpp:72]: vpdStartOffset

File: vpd-parser/parser_factory.cpp [hunk 8]
```diff
@@ -24,15 +24,16 @@ namespace factory
 {
 interface::ParserInterface*
     ParserFactory::getParser(const Binary& vpdVector,
-                             const std::string& inventoryPath)
+                             const std::string& inventoryPath,
+                             const std::string& file, uint32_t offset)
```
Comment (jinuthomas) [inline vpd-parser/parser_factory.cpp:28]: use understandable variable names as mentioned before

File: vpd-parser/parser_factory.hpp [hunk 9]
```diff
@@ -34,7 +34,8 @@ class ParserFactory
      * @return - Pointer to concrete parser class object.
```
Comment (jinuthomas) [inline vpd-parser/parser_factory.hpp:34]: Params missing

File: vpd_tool_impl.cpp [hunk 10]
```diff
@@ -51,7 +51,8 @@ static void
     Binary vpdVector{};
 
     vpdVector = getVpdDataInVector(js, constants::systemVpdFilePath);
-    ParserInterface* parser = ParserFactory::getParser(vpdVector, invPath);
+    ParserInterface* parser = ParserFactory::getParser(vpdVector, invPath,
+        constants::systemVpdFilePath, 0);
```
Comment (jinuthomas) [inline vpd_tool_impl.cpp:55]: NO magic numbers please

File: vpd_tool_impl.cpp [hunk 11]
```diff
@@ -579,7 +580,16 @@ void VpdTool::readKwFromHw(const uint32_t& startOffset)
     const std::string& inventoryPath =
         jsonFile["frus"][fruPath][0]["inventoryPath"];
 
-    Impl obj(completeVPDFile, (constants::pimPath + inventoryPath));
+    uint32_t offset = 0;
```
Comment (jinuthomas) [inline vpd_tool_impl.cpp:583]: vpdStartOffset

**Review by jinuthomas — CHANGES_REQUESTED** [review]

**Comment by SunnySrivastava1984** [comment]
@GiridhariKrishnan As discussed, Please push the PR on Gerrit.

**Review by jinuthomas — COMMENTED** [review]
Looks Good to me, just one question
If vpdFileStream.open fails 
what will vpdFileStream.seek give ? Do we need to check if it is open before trying a seek ?

**Review by jinuthomas — APPROVED** [review]

**Pull request merged**